### PR TITLE
New case of create_large_raw_img

### PR DESCRIPTION
**Pull request by hellohellenmao**
Test creating large raw images on different file systems.

ID: 1801538

Signed-off-by: Tingting Mao <timao@redhat.com>

### CODE REVIEW

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.1.x86_64.io-github-autotest-qemu.create_large_raw_img: PASS (3.76 s)

**Comment by hellohellenmao** [comment]
@lolyu @YongxueHong Help to take a review, please.

**Comment by lolyu** [comment]
what is `[NEED UPDATE BY ME]`?

**Comment by hellohellenmao** [comment]
> 1. Using `loop device` is not a test step listed in Polarion.
I usually use 'loop device' in my daily test, and I think 'loop devices' are the easiest way to set up.
> 2. how do you guarantee the assumption that `{data_dir}/images/` is `xfs`-formatted?
Updated.
> 3. it would be much better to split `xfs` and `ext4` into config variants.
Update, thanks.

File: qemu/tests/cfg/create_large_raw_img.cfg [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    start_vm = no
+    create_image = no
+    remove_image = yes
+    images = "filename loop"
+    image_name_filename = "images/filename"
+    image_name_loop = "images/loop"
+    image_size_filename = 10240000T
+    image_size_loop = 50G
+    image_format_loop = raw
+    file_systems = "xfs ext4"
```
Comment (lolyu) [inline qemu/tests/cfg/create_large_raw_img.cfg]: better to:
```
variants:
    - over_xfs:
        ...
    - over_ext4:
        ...
```

File: qemu/tests/create_large_raw_img.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    try_create_img(filename, xfs)
+    process.run("umount %s" % root_dir)
+
+    logging.info("Test creating an image with large size on EXT4.")
+    setup_file_system_env(new_dev, ext4, root_dir)
+    params["image_size_filename"] = "16T"
+    filename = QemuImg(params.object_params(filename_img), root_dir,
+                       filename_img)
+    try_create_img(filename, ext4)
+
```
Comment (lolyu) [inline qemu/tests/create_large_raw_img.py]: if you split the config into variants, code from LINE 91 to 96 could be removed.

File: qemu/tests/cfg/create_large_raw_img.cfg (hunk 1 above)
Comment (hellohellenmao) [reply qemu/tests/cfg/create_large_raw_img.cfg]: Updated, thanks

File: qemu/tests/create_large_raw_img.py (hunk 2 above)
Comment (hellohellenmao) [reply qemu/tests/create_large_raw_img.py]: Updated, thanks

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_xfs: PASS (2.82 s)
 (2/2) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_ext4: PASS (3.54 s)

File: qemu/tests/cfg/create_large_raw_img.cfg [hunk 3]
```diff
@@ -0,0 +1,21 @@
+- create_large_raw_img:
+    only raw
+    virt_test_type = qemu
+    type = create_large_raw_img
+    start_vm = no
+    create_image = no
+    remove_image = yes
+    images = "filename loop"
```
Comment (lolyu) [inline qemu/tests/cfg/create_large_raw_img.cfg]: `filename` is misleading.

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_xfs: PASS (3.97 s)
 (2/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_ext4: PASS (3.20 s)

File: qemu/tests/cfg/create_large_raw_img.cfg (hunk 3 above)
Comment (hellohellenmao) [reply qemu/tests/cfg/create_large_raw_img.cfg]: Maybe you could list all of advises in one time, please. Or I have to push and re-test the case for a few times.

**Comment by hellohellenmao** [comment]
@lolyu @YongxueHong Help to review, please.

File: qemu/tests/cfg/create_large_raw_img.cfg [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    images = "large loop"
+    image_name_large = "images/large"
+    image_name_loop = "images/loop"
+    image_size_loop = 50G
+    image_format_loop = raw
+    variants:
+      - over_xfs:
+        file_sys = "xfs"
+        image_size_large = 10240000T
+        err_info = "Image size must be less than 8 EiB!"
```
Comment (YongxueHong) [inline qemu/tests/cfg/create_large_raw_img.cfg]: Should be 
```
      - over_xfs:
            file_sys = "xfs"
            image_size_large = 10240000T
            err_info = "Image size must be less than 8 EiB!"
```

Comment (hellohellenmao) [reply qemu/tests/cfg/create_large_raw_img.cfg]: Updated, thanks.

Comment (hellohellenmao) [reply qemu/tests/cfg/create_large_raw_img.cfg]: Review again, please.

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_xfs: PASS (3.36 s)
 (2/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_ext4: PASS (3.31 s)

**Comment by lolyu** [comment]
There is no need to create a `loop device`, you could simply format a file with `mkfs` then mount it as a `loopback`.
this is already implemented in `avocado.utils.partition`
```
# python
>>> from avocado.utils import partition as p
>>> p0 = p.Partition('/home/disk.img', loop_size=2048, mountpoint="/mnt")
>>> p0.mkfs("xfs")
>>> p0.mount()
# mount | tail -n1
/home/disk.img on /mnt type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
```

File: qemu/tests/create_large_raw_img.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import logging
+
+from avocado.core import exceptions
```
Comment (lolyu) [inline qemu/tests/create_large_raw_img.py]: `from avocado import TestError`

File: qemu/tests/create_large_raw_img.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+from avocado.utils import process
+from avocado.utils import partition as p
```
Comment (lolyu) [inline qemu/tests/create_large_raw_img.py]: better to remove the rename as `p`.

File: qemu/tests/create_large_raw_img.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    3. Create a raw image with large size on the EXT4 file system and
+    check the output info.
+
+    :param test: VT test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+    large_img, loop_img = params["images"].split()
+    root_dir = data_dir.get_data_dir()
+    loop = QemuImg(params.object_params(loop_img), root_dir, loop_img)
```
Comment (lolyu) [inline qemu/tests/create_large_raw_img.py]: why define a `loop` image here?

File: qemu/tests/create_large_raw_img.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    loop_filename = loop.image_filename
+    loop_size = loop.image_size
+    file_sys = params["file_sys"]
+
+    root_dir = root_dir + "/tmp"
```
Comment (lolyu) [inline qemu/tests/create_large_raw_img.py]: `mnt_dir = os.path.join(root_dir, "tmp")`
1. please use a different name.
2. please use `os.path.join`

File: qemu/tests/create_large_raw_img.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    process.run("mkdir %s" % root_dir)
```
Comment (lolyu) [inline qemu/tests/create_large_raw_img.py]: `os.mkdir`

File: qemu/tests/create_large_raw_img.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    try:
+        large.create(large.params)
+    except exceptions.TestError as err:
+        if params["err_info"] not in str(err):
+            test.fail("CML failed with unexpected output: %s" % err)
+    else:
+        test.fail("There is no error when creating an image with large size.")
+
+    part.unmount()
+    process.run("rm -rf %s" % root_dir)
```
Comment (lolyu) [inline qemu/tests/create_large_raw_img.py]: `os.rmdir`

File: qemu/tests/create_large_raw_img.py (hunk 9 above)
Comment (hellohellenmao) [reply qemu/tests/create_large_raw_img.py]: Thanks.

File: qemu/tests/create_large_raw_img.py (hunk 7 above)
Comment (hellohellenmao) [reply qemu/tests/create_large_raw_img.py]: Updated, thanks.

File: qemu/tests/create_large_raw_img.py (hunk 6 above)
Comment (hellohellenmao) [reply qemu/tests/create_large_raw_img.py]: Why `p` is better?

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_xfs: PASS (412.09 s)
 (2/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_ext4: PASS (420.11 s)

**Comment by hellohellenmao** [comment]
@lolyu Help to take a review, please.

**Comment by lolyu** [comment]
test under python2 please.

File: qemu/tests/cfg/create_large_raw_img.cfg [hunk 11]
```diff
@@ -0,0 +1,18 @@
+- create_large_raw_img:
+    only raw
+    virt_test_type = qemu
+    type = create_large_raw_img
+    start_vm = no
+    create_image = no
+    images = "large"
+    image_name_large = "images/large"
```
Comment (lolyu) [inline qemu/tests/cfg/create_large_raw_img.cfg]: `image_name_large = "large"`, no need to create a subdirectory under the mount dir.

File: qemu/tests/cfg/create_large_raw_img.cfg [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    image_size_loop = 51200
```
Comment (lolyu) [inline qemu/tests/cfg/create_large_raw_img.cfg]: better to name it `loop_file_size`.

Comment (hellohellenmao) [reply qemu/tests/cfg/create_large_raw_img.cfg]: Updated, thanks.

**Comment by hellohellenmao** [comment]
(1/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_xfs: PASS (409.00 s)
 (2/2) Host_RHEL.m8.u2.product_av.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_ext4: PASS (418.63 s)

**Comment by hellohellenmao** [comment]
@lolyu Help to review, please.

**Comment by hellohellenmao** [comment]
@lolyu Help to review, please.

**Comment by hellohellenmao** [comment]
The result in python2:
 (1/2) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_xfs: PASS (347.82 s)
 (2/2) Host_RHEL.m7.u8.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.create_large_raw_img.over_ext4: PASS (350.70 s)

**Pull request merged**